Extract method from RBR.adjustRegions(Vertex sw)

matheus220 opened the pull request:
Extract ArrayList<Region> adjustRegion(Region r) from RBR.adjustRegions(Vertex sw)

Commit: Ref

TaskNumber: #70

matheus220 committed:
Extract method from RBR.adjustRegions(Vertex sw)

Extract ArrayList<Region> adjustedRegionsFrom(Region r) from RBR.adjustRegions(Vertex sw)

Commit: Ref

TaskNumber: #70

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -134,22 +134,28 @@ private void adjustRegions(Vertex sw) {
+    private ArrayList<Region> adjustRegions(Region r) {
+        ArrayList<Region> newRegions = new ArrayList<>();
+        Set<Vertex> outsiders = r.outsiders();
+        assert (!outsiders.isEmpty()) : "Outsiders is Empty";
```
You shouldn't assert that. 'outsiders' can be empty sometimes and it means that you don't need to adjust that region 'r'. You could in that case just add 'r' to 'newRegions' and return it.

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -134,22 +134,28 @@ private void adjustRegions(Vertex sw) {
-            newRegions.addAll(makeRegions(trulyDestinationsInOutsidersRange, currentRegion.inputPorts(), currentRegion.outputPorts()));
         }
         regionsForVertex.get(sw).removeAll(regionsToBeRemoved);
         regionsForVertex.get(sw).addAll(newRegions);
```
The way that you made the extraction 'regionsToBeRemoved' will contain all sw's regions (therefore the statement at 140 will empty the list 'regionsForVertex.get(sw)') and 'newRegions' will contain all regions to be used thereafter. Instead of emptying the list and then filling it completely you could replace the whole list with only one statement. That way you'll no more need 'regionsToBeRemoved'.

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -134,22 +134,28 @@ private void adjustRegions(Vertex sw) {
         regionsForVertex.get(sw).addAll(newRegions);
     }
 
+    private ArrayList<Region> adjustRegions(Region r) {
```
As you are 'getting' something, a better name to that method could be 'adjustedRegions' or 'adjustedRegionsFrom'.

jmferreiratech commented:
I'm considering assigning you #74 to integrate that patch. What do you think?

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -132,22 +132,26 @@ public void regionsComputation() {
-            Set<Vertex> outsiders = currentRegion.outsiders();
-            if(outsiders.isEmpty())
-                continue;
+        for (Region currentRegion : regionsForVertex.get(sw))
```
User curly braces here.

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -132,22 +132,26 @@ public void regionsComputation() {
+    }
+
+    private ArrayList<Region> adjustedRegionsFrom(Region r) {
+        ArrayList<Region> newRegions = new ArrayList<>();
```
Rename it to 'adjustedRegions'.

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -132,22 +132,26 @@ public void regionsComputation() {
+        Set<Vertex> outsiders = r.outsiders();
+        if(outsiders.isEmpty()){
+            newRegions.add(r);
+        } else{
```
Put a space before the opening curly brace.

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -132,22 +132,26 @@ public void regionsComputation() {
-            ArrayList<Region> regionsToAdd = splitRegionExcludingOutsiders(currentRegion, outsidersBox);
+            ArrayList<Region> regionsToAdd = splitRegionExcludingOutsiders(r, outsidersBox);
             newRegions.addAll(regionsToAdd);
             // use others routers to make others regions
```
You can remove that commented line.

matheus220 commented:
In relation the issue #74. It's ok.

jmferreiratech commented on src/rbr/RBR.java:
```diff
@@ -131,22 +131,26 @@ public void regionsComputation() {
+    }
+
+    private ArrayList<Region> adjustedRegions(Region r) {
+        ArrayList<Region> newRegions = new ArrayList<>();
```
I'm sorry. I didn't make myself clear enough. 'adjustedRegionsFrom' was a very good name to that method. What I asked to change was the variable 'newRegions' which I think could be named 'adjustedRegions'

rmilfont reviewed (approved):
Now, just do the #74 and it will be merged.

matheus220 committed:
Unit test RBR.adjustedRegionsFrom()

Add unit tests to RBR.adjustedRegionsFrom()

Commit: Add

TaskNumber: #74

jmferreiratech commented on tests/rbr/AdjustRegionsTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    public void checkIP() throws Exception {
```
Name your tests after what you are trying to prove.
I'd name this one 'adjustedRegionsHaveSameInputPortsAsOriginal'.

jmferreiratech commented on tests/rbr/AdjustRegionsTest.java:
```diff
@@ -0,0 +1,59 @@
+        Assert.assertEquals(region.destinations().size(), contDst);
+    }
+
+}
```
Put a newline at the end of file.

jmferreiratech commented on tests/rbr/AdjustRegionsTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    public void checkNoOutsiders (){
```
Shouldn't all tests throw exceptions?

jmferreiratech commented on tests/rbr/AdjustRegionsTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    public void checkDestinations (){
```
Is this test passing? I think it shouldn't. I'd rather assert that the union of the destinations of adjusted regions is equal to the set of destinations of the original region.

jmferreiratech commented on tests/rbr/AdjustRegionsTest.java:
```diff
@@ -0,0 +1,59 @@
+    private Set<Character> ip = new HashSet<>();
+    private Set<Character> op = new HashSet<>();
+    private Set<Vertex> dst = new HashSet<>();
+    private Region region;
```
'region' is the only attribute that you use outside 'setUp()' and so should be the only attribute, don't you think so?

jmferreiratech commented on tests/rbr/AdjustRegionsTest.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Test
+    public void checkIP() throws Exception {
+        for(Region r : adjustedRegionsFrom(region)){
```
I'd name 'r' as 'adjustedRegion' and 'region' as 'original'.

jmferreiratech reviewed (changes requested):
I'd add a 'RegionWithoutOutsidersIsntAdjusted' test.

What should happen if you pass a region without destinations?

You could test also that no adjusted region contains an outsider.

matheus220 commented on src/rbr/RBR.java:
```diff
@@ -129,24 +129,28 @@ public void regionsComputation() {
-            newRegions.addAll(makeRegions(trulyDestinationsInOutsidersRange, currentRegion.inputPorts(), currentRegion.outputPorts()));
+            List<Region> regionsToAdd = splitRegionExcludingOutsiders(r, outsidersBox);
+            adjustedRegions.addAll(regionsToAdd);
+            adjustedRegions.addAll(makeRegions(trulyDestinationsInOutsidersRange, r.inputPorts(), r.outputPorts()));
```
Why not just leave line 151 inside the else? I think still adjusting regions correctly.

```
} else {
    adjustedRegions.addAll(makeRegions(r.destinations(), r.inputPorts(), r.outputPorts()));
}
```

jmferreiratech replied:
I actually would prefer no if/else at all here. Although it can be a next step (as would require 'fix' 'splitRegionExcludingOutsiders' and 'makeRegions') is a good practice to minimize things inside conditionals, so go ahead.

matheus220 commented:
When we pass a region without destinations an error occurred in Region.outsiders() when it calls Region.box(), because box.min and box.max will be null. So the test regionWithoutDestinationsIsntAdjusted was not done.

jmferreiratech commented:
@matheus220 please implement the regionWithoutDestinationsIsntAdjusted test and annotate it with "@Ignore @Test".
But tell me, which part of the code fails to handle the null min/max? So that we can fix it.

matheus220 commented:
The problem was in Region.java, line 61. I put an if, what do you think?

jmferreiratech commented:
That's ok, but I'd prefer to make sure it'll be properly fixed.
Could you please open an issue requesting that?

jmferreiratech commented on tests/rbr/AdjustRegionsTest.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Test
+    public void adjustedRegionsWithoutOutsiders() throws Exception {
```
Don't you think that 'adjustedRegionsHaveNoOutsider' would reveal more the intent of this test?

jmferreiratech replied:
I still think you can add a test showing that each 'adjustedRegion' does not contains any 'original.outsiders()'. I'd call it 'adjustedRegionsDoNotIncludeOriginalOutsiders'.

matheus220 commented:
Ok. I opened an issue, please analyze it. And made the requested modifications.

jmferreiratech commented:
@matheus220, it seems that with your patch you could remove RBR.splitRegionExcludingOutsiders() and Region.destinationsIn() methods. Could you check that?

matheus220 commented:
Ok, I'll check

matheus220 commented:
I checked, the main difference of the new path is that the new regions will be disjoint, but this does not affect the operation of the system. I did several tests and all were ok. So I think we can remove the RBR.splitRegionExcludingOutsiders() and Region.destinationsIn() methods.

jmferreiratech commented:
Fine. So, please remove them and I'll merge it.

matheus220 committed:
Remove methods

By modifying the method RBR.adjustedRegionsFrom() the methods RBR.splitRegionExcludingOutsiders() and Region.destinationsIn() are not used. Then they were removed.

Commit:Rem

TaskNumber:#70

matheus220 commented:
Ready. With that the issue #71 can be removed.
